File: src/mrl/creatures/Creature.java
/*
 *  "MysteriesRL", Java Roguelike game.
 *  Copyright (C) 2015 by Serg V. Zhdanovskih (aka Alchemist, aka Norseman).
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package mrl.creatures;

import jzrlib.core.CreatureEntity;
import jzrlib.core.GameSpace;
import jzrlib.map.IMap;
import mrl.core.ITerminalEntity;
import mrl.game.Game;
import mrl.maps.Layer;

/**
 *
 * @author Serg V. Zhdanovskih
 */
public abstract class Creature extends CreatureEntity implements ITerminalEntity
{
    protected int fHPCur;

    private boolean fIsPlayer;

    public Creature(GameSpace space, Object owner)
    {
        super(space, owner);
    }

    @Override
    public final Game getSpace()
    {
        return (Game) this.fSpace;
    }
    
    public final IMap getMap()
    {
        return (IMap) this.Owner;
    }

    public final void setMap(IMap value)
    {
        IMap prevMap = this.getMap();
        if (prevMap != null) { 
            ((Layer) prevMap).getCreatures().remove(this);
        }

        this.Owner = value;

        if (value != null) {
            ((Layer) value).getCreatures().add(this);
        }
    }
    
    public void ascend()
    {
        IMap map = this.getMap();
        if (map == this.getSpace().getPlainMap()) {
            //
        }
        if (map == this.getSpace().getCellarsMap()) {
            this.setMap(this.getSpace().getPlainMap());
        }
        if (map == this.getSpace().getDungeonsMap()) {
            this.setMap(this.getSpace().getCellarsMap());
        }
    }
    
    public void descend()
    {
        IMap map = this.getMap();
        if (map == this.getSpace().getPlainMap()) {
            this.setMap(this.getSpace().getCellarsMap());
        }
        if (map == this.getSpace().getCellarsMap()) {
            this.setMap(this.getSpace().getDungeonsMap());
        }
        if (map == this.getSpace().getDungeonsMap()) {
            //
        }
    }

    public final boolean isPlayer()
    {
        return this.fIsPlayer;
    }

    public final void setIsPlayer(boolean value)
    {
        this.fIsPlayer = value;
    }
    
    public void doTurn()
    {
        // dummy
    }

    @Override
    public void moveTo(int newX, int newY)
    {
        super.moveTo(newX, newY);
    }
    
    // <editor-fold defaultstate="collapsed" desc="Combat implementation">

    public final int getHP()
    {
        return this.fHPCur;
    }

    public final void setHP(int hp)
    {
        this.fHPCur = hp;

        if (this.fHPCur > this.getHPMax()) {
            this.fHPCur = this.getHPMax();
        }
    }

    public final boolean isAlive()
    {
        return this.fHPCur > 0;
    }

    public int getHPMax()
    {
        return 0;
    }

    public void takeDamage(Damage damage)
    {
        this.fHPCur = this.fHPCur - damage.Value;

        if (!this.isAlive()) {
            this.die(damage.getInflictor());
        }
    }

    public void die(Creature killer)
    {
    }

    public void kill(Creature victim)
    {
    }
    
    public int getDamageValue()
    {
        return 0;
    }

    public int getDamageType()
    {
        return Damage.DMG_GENERIC;
    }

    protected void attack(Creature enemy)
    {
    }

    public int getEquipBonus(String bonusId)
    {
        return 0;
    }
    
    /*
     * This method inflicts damage on given entity,
     * based on current combat mechanics of this entity
     */
    public void inflictDamage(Creature enemy)
    {
        if (enemy == null || enemy == this) {
            return;
        }

        this.attack(enemy);

        int type = this.getDamageType();
        int value = this.getDamageValue();

        // non-lethal weapons ignore all damage including stat bonuses
        if (type == Damage.DMG_NONLETHAL) {
            value = 0;
        }

        Damage dmg = new Damage(value, type, this);
        enemy.takeDamage(dmg);
    }

    public abstract int getFovRadius();
    
    public abstract int getHearRadius();
    
    // </editor-fold>
}

File: project/Creatures/Creature.cs
/*
 *  "PrimevalRL", roguelike game.
 *  Copyright (C) 2015, 2017 by Serg V. Zhdanovskih.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using PrimevalRL.Game;
using PrimevalRL.Maps;
using ZRLib.Core;
using ZRLib.Map;

namespace PrimevalRL.Creatures
{
    public abstract class Creature : CreatureEntity, ITerminalEntity
    {
        public abstract int AppearanceColor { get; }
        public abstract char Appearance { get; }
        protected int fHPCur;

        private bool fIsPlayer;

        protected Creature(GameSpace space, object owner)
            : base(space, owner)
        {
        }

        public new MRLGame Space
        {
            get { return (MRLGame)fSpace; }
        }

        public IMap Map
        {
            get {
                return (IMap)Owner;
            }
            set {
                IMap prevMap = Map;
                if (prevMap != null) {
                    ((Layer)prevMap).Creatures.Remove(this);
                }
    
                Owner = value;
    
                if (value != null) {
                    ((Layer)value).Creatures.Add(this);
                }
            }
        }

        public bool Player
        {
            get {
                return fIsPlayer;
            }
        }

        public bool IsPlayer
        {
            set {
                fIsPlayer = value;
            }
        }

        public virtual void DoTurn()
        {
            // dummy
        }

        public override void MoveTo(int newX, int newY)
        {
            base.MoveTo(newX, newY);
        }

        public int HP
        {
            get {
                return fHPCur;
            }
            set {
                fHPCur = value;
    
                if (fHPCur > HPMax) {
                    fHPCur = HPMax;
                }
            }
        }

        public bool Alive
        {
            get {
                return fHPCur > 0;
            }
        }

        public virtual int HPMax
        {
            get {
                return 0;
            }
        }

        public virtual void TakeDamage(Damage damage)
        {
            fHPCur = fHPCur - damage.Value;

            if (!Alive) {
                Die(damage.Inflictor);
            }
        }

        public virtual void Die(Creature killer)
        {
        }

        public virtual void Kill(Creature victim)
        {
        }

        public virtual int DamageValue
        {
            get {
                return 0;
            }
        }

        public virtual int DamageType
        {
            get {
                return Damage.DMG_GENERIC;
            }
        }

        protected virtual void Attack(Creature enemy)
        {
        }

        public virtual int GetEquipBonus(string bonusId)
        {
            return 0;
        }

        /*
         * This method inflicts damage on given entity,
         * based on current combat mechanics of this entity
         */
        public virtual void InflictDamage(Creature enemy)
        {
            if (enemy == null || enemy == this) {
                return;
            }

            Attack(enemy);

            int type = DamageType;
            int value = DamageValue;

            // non-lethal weapons ignore all damage including stat bonuses
            if (type == Damage.DMG_NONLETHAL) {
                value = 0;
            }

            Damage dmg = new Damage(value, type, this);
            enemy.TakeDamage(dmg);
        }

        public abstract int FovRadius { get; }

        public abstract int HearRadius { get; }
    }
}
